backend/data_fetcher.py

```diff
@@ -0,0 +1,168 @@
+            Example:
+                [
+                    {
+                        'name':'sys',
+                        'data':[
+                            [time,power],
+                            [time,power]
+                        ]},
+                    {
+                        'name': 'channel2',
+                        'data': [
+                            [time,power]
+                        ]
+                    }
+                ]
+        """
+        self._metadata = Metadata(
+            self._preprocess_dir, bucket=self._preprocess_bucket)
+        self._metadata.load()
+
+        if timespan_start is None:
+            timespan_start = self._metadata['start']
+        if timespan_end is None:
+            timespan_end = self._metadata['end']
+
+        if timespan_start > self._metadata['end'] or timespan_end < self._metadata['start']:
+            return []
+
+        required_frequency = number_records / (timespan_end - timespan_start)
+
+        # Finds Downsample Level.
+        target_level_index = self._binary_search(
+            [self._metadata['levels'][level_name]['frequency']
+             for level_name in self._metadata['levels']['names']],
+            required_frequency, True)
+
+        target_level = self._metadata['levels'][self._metadata['levels']
+                                                ['names'][target_level_index]]
+
+        level_metadata = Metadata(
+            self._preprocess_dir, self._preprocess_bucket, strategy, utils.get_level_name(
+                target_level_index))
+        level_metadata.load()
+        first_slice = self._binary_search([level_metadata[single_slice]
+                                           for single_slice in target_level['names']],
+                                          timespan_start)
+        last_slice = self._binary_search([level_metadata[single_slice]
+                                          for single_slice in target_level['names']],
+                                         timespan_end)
+        target_slices_names = target_level['names'][first_slice:last_slice+1]
+        target_slice_paths = [utils.get_slice_path(
+            self._preprocess_dir,
+            utils.get_level_name(target_level_index),
+            single_slice, strategy) for single_slice in target_slices_names]
+
+        # Reads records and downsamples.
+        target_slices = LevelSlicesReader(
+            target_slice_paths, self._preprocess_bucket)
+        target_slices.read(timespan_start, timespan_end)
+        number_target_records = target_slices.get_records_count()
+
+        target_slices.downsample(strategy, max_records=number_records)
+        downsampled_data = target_slices.format_response()
+        number_result_records = target_slices.get_records_count()
+
+        if number_target_records == 0:
+            precision = 0
+        else:
+            precision = number_result_records / \
```

Comment:
It could be a lot work... 
This calculation involves reading overall metadata and level metadata as well as number of records.
I'm afraid there's not enough time to do a runnable mock object given that the calculation is at the end of the data fetch...